akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala

```diff
@@ -302,6 +302,28 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         val error = entityDataIn.expectError()
         error.getMessage shouldBe "Stream with ID [1] was closed by peer with code INTERNAL_ERROR(0x02)"
       }
+
+      // https://github.com/akka/akka-http/issues/2957
+      "close the stream when we receive a RST after we have half-closed ourselves as well" in new WaitingForRequestData {
+        import akka.http.scaladsl.client.RequestBuilding._
+        // Client sends the request, but doesn't close the stream yet. This is a bit weird, but it's whet grpcurl does ;)
+        val post = Post("/grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo")
+        sendHEADERS(streamId = 1, endStream = false, endHeaders = true, encodeRequestHeaders(request))
+        sendDATA(streamId = 1, endStream = false, ByteString(0, 0, 0, 0, 0x10, 0x22, 0x0e) ++ ByteString.fromString("GreeterService"))
+
+        // We emit a 404 response, half-closing the stream.
+        emitResponse(streamId = 1, HttpResponse(StatusCodes.NotFound))
+
+        // The client closes the stream with a protocol error. We should look into whether this is legit - but
+        // in any case we should handle it correctly.
+        sendRST_STREAM(streamId = 1, ErrorCode.PROTOCOL_ERROR)
+        entityDataIn.expectError()
+
+        // Unfortunately we can't check the problem in
```

Comment:
So, before the fix the connection is just aborted? Could we check that?

Reply:
Maybe the warning should have been a crash instead?

Reply:
The behavior before and after is the same: before it'd abort the incoming request stream (OK), and try to close the outgoing response stream, but the latter would fail since it was already closed. Crashing everything just because one stream is in an unexpected state seems a bit dangerous...

Reply:
Any reason not to use the regular `EventFilter` here?

Reply:
hmm, can you use EventFilter to check we do _not_ get a log message?

Reply:
Good point, maybe with `occurrences=0` and a bit of sleep in the intercept?